src/autodiff_model.jl

```diff
@@ -52,18 +52,14 @@ end
 function ADNLPModel(f, x0::AbstractVector; y0::AbstractVector = eltype(x0)[],
                     lvar::AbstractVector = eltype(x0)[], uvar::AbstractVector = eltype(x0)[],
                     lcon::AbstractVector = eltype(x0)[], ucon::AbstractVector = eltype(x0)[],
-                    c = (args...)->throw(NotImplementedError("cons")),
+                    c = nothing,
```

Comment:
How is this going to work if the user calls `cons()`?

Reply:
`ERROR: MethodError: objects of type Nothing are not callable`

Reply:
I'm not really convinced by this. I found the NotImplementedError helpful. In the blog, the author makes an error (passing a vector instead of a matrix) and implemented an error message that isn't helpful. I get the gist but in this particular case, does `ERROR: MethodError: objects of type Nothing are not callable` really help understand what's wrong?

Reply:
Maybe for this specific model, having `NotImplemented` may be helpful, but the overall change is for the API. `NotImplemented` would say `foo not implemented` and the current way says something like `no method matching obj(::DummyModel, ::Array{Float64,1})` and it shows the closest candidates.

Reply:
Could we use something like `c = () -> nothing` so as to trigger a MethodError? Or `c = function cons end`?

Reply:
Sure, I'll think of something.

Reply:
I think the most descriptive way is to create two models `ADNLPUncModel` and `ADNLPConModel`, both `<: ADNLPModels <: AbstractNLPModel`, and then change the functions to receive either `ADNLPModels` or `ADNLPConModel`.

Reply:
Why not consider that unconstrained models simply provide 0 as constraints? Of course then one has to ensure that when not empty,  lcon <= 0 <= ucon. Perhaps two constructors achieve that.
Imagine benchmarking a solver for constrained problems, and including some unconstrained instances in the testset, the solver would crash because there is no constraint?

Reply:
Actually that's a great point. We do define lvar, lcon, etc as empty arrays. So c should also return an empty array.

Reply:
Also there's an inconsistency with the lvar/uvar here and https://github.com/JuliaSmoothOptimizers/NLPModels.jl/blob/master/src/nlp_types.jl#L101

Reply:
Thanks, that's much better. I've added a new commit changing that and defining `lvar` and `uvar` correctly on the keyword default values.

Reply:
Thanks!